core/matrix/fbcsr.cpp

```diff
@@ -0,0 +1,504 @@
+{
+    result->values_ = this->values_;
+    result->col_idxs_ = this->col_idxs_;
+    result->row_ptrs_ = this->row_ptrs_;
+    result->set_size(this->get_size());
+    result->bs_ = this->bs_;
+    result->nbcols_ = this->nbcols_;
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::move_to(
+    Fbcsr<next_precision<ValueType>, IndexType> *const result)
+{
+    this->convert_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::convert_to(
+    Dense<ValueType> *const result) const
+{
+    auto exec = this->get_executor();
+    auto tmp = Dense<ValueType>::create(exec, this->get_size());
+    exec->run(fbcsr::make_convert_to_dense(this, tmp.get()));
+    tmp->move_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::move_to(Dense<ValueType> *const result)
+{
+    this->convert_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::convert_to(
+    Csr<ValueType, IndexType> *const result) const
+{
+    auto exec = this->get_executor();
+    auto tmp = Csr<ValueType, IndexType>::create(
+        exec, this->get_size(), this->get_num_stored_elements(),
+        result->get_strategy());
+    exec->run(fbcsr::make_convert_to_csr(this, tmp.get()));
+    tmp->move_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::move_to(
+    Csr<ValueType, IndexType> *const result)
+{
+    this->convert_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::convert_to(
+    SparsityCsr<ValueType, IndexType> *const result) const
+{
+    auto exec = this->get_executor();
+    auto tmp = SparsityCsr<ValueType, IndexType>::create(
+        exec,
+        gko::dim<2>{static_cast<size_type>(this->get_num_block_rows()),
+                    static_cast<size_type>(this->get_num_block_cols())},
+        this->get_num_stored_blocks());
+
+    tmp->col_idxs_ = this->col_idxs_;
+    tmp->row_ptrs_ = this->row_ptrs_;
+    tmp->value_ = Array<ValueType>(exec, {one<ValueType>()});
+    tmp->move_to(result);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::move_to(
+    SparsityCsr<ValueType, IndexType> *const result)
+{
+    this->convert_to(result);
+}
+
+
+/*
+ * Currently, this implementation is sequential and has complexity
+ * O(nnz log(nnz)).
+ * @note Can this be changed to a parallel O(nnz) implementation?
+ */
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::read(const mat_data &data)
+{
+    GKO_ENSURE_IN_BOUNDS(data.nonzeros.size(),
+                         std::numeric_limits<index_type>::max());
+
+    const auto nnz = static_cast<index_type>(data.nonzeros.size());
+    const int bs = this->bs_;
+
+    using Block_t = detail::DenseBlock<value_type>;
```

Comment:
This means a single allocation for each block of the matrix, this might be really inefficient. An alternative using only a single allocation would be building a map `(row,col)->block_idx`, allocating and in a second pass filling the blocks.

Reply:
Yes, true. But (1) will that be really inefficient? Looking up `std::map` is not exactly instantaneous, AFAIK, and we're also reading the block for each allocation. Typically block sizes would be 3x3 to 7x7 or so. So my assumption was that time spent in allocation may not be that much, relative to the whole thing. I'm not sure a 2-pass approach would actually be faster, though indeed it might be. (2) This function will typically only be called once per program execution. Do you think it's worth going and changing it?

Reply:
See this.